From 0a81fcb4c61d154c8db1ba9a9081e9651b4b99bc Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 8 Sep 2014 13:07:49 -0400
Subject: [tools] perf/report/tui: Fix a bug when --fields/sort is given

Message-id: <1410181752-25631-225-git-send-email-jolsa@redhat.com>
Patchwork-id: 91974
O-Subject: [PATCH RHEL7.1 BZ1134356 224/307] perf report/tui: Fix a bug when --fields/sort is given
Bugzilla: 1134356
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1134356
https://bugzilla.redhat.com/show_bug.cgi?id=1134356

upstream
========
commit c0f1527b7e004f9a91e488f05c251213d16ad7ac
Author: Namhyung Kim <namhyung@kernel.org>
Date: Wed Apr 16 11:16:33 2014 +0900

description
===========
The hists__filter_entries() function is called when down arrow key is
pressed for navigating through the entries in TUI.  It has a check for
filtering out entries that have very small overhead (under min_pcnt).

However it just assumed the entries are sorted by the overhead so when
it saw such a small overheaded entry, it just stopped navigating as an
optimization.  But it's not true anymore due to new --fields and
--sort optoin behavior and this case users cannot go down to a next
entry if ther's an entry with small overhead in-between.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/ui/browsers/hists.c b/tools/perf/ui/browsers/hists.c
index 92d128f..169224c 100644
--- a/tools/perf/ui/browsers/hists.c
+++ b/tools/perf/ui/browsers/hists.c
@@ -812,10 +812,7 @@ static struct rb_node *hists__filter_entries(struct rb_node *nd,
   if (total)
    percent = h->stat.period * 100.0 / total;
 
-  if (percent < min_pcnt)
-   return NULL;
-
-  if (!h->filtered)
+  if (!h->filtered && percent >= min_pcnt)
    return nd;
 
   nd = rb_next(nd);
-- 
1.7.1